elastic/beats
# Add complete k8s metadata through composable provider

## What does this PR do?
Add all k8s metadata via the provider.

## Why is it important?
To support full metadata access like in Beats.

## Checklist




## TODOs

- [x] Verify that short-living pods are handled properly.

## How to test this PR locally
1. Enable a static autodiscovery rule in Agent's config:
```
providers.kubernetes:
  scope: node
  kube_config: /Users/chrismark/.kube/config
  node: "kind-control-plane"
  cleanup_timeout: 360s
  resources:
    pod:
      enabled: true

inputs:
- name: redis
  type: redis/metrics
  use_output: default
  meta:
    package:
      name: redis
      version: 0.3.6
  data_stream:
    namespace: default
  streams:
    - data_stream:
        dataset: redis.info
        type: metrics
      metricsets:
        - info
      hosts:
        - '${kubernetes.pod.ip}:${kubernetes.container.port}'
      idle_timeout: 20s
      maxconn: 10
      network: tcp
      period: 10s
      condition: ${kubernetes.labels.app} == 'redis' AND ${kubernetes.container.port_name} == 'web'
```
2. Deploy a Redis pod that triggers the condition:
```
apiVersion: v1
kind: Pod
metadata:
  name: redis
  labels:
    role: main
    app: redis
  annotations:
    # this had no meaning
    co.elastic.metrics.redis.XYZ/module: redis
spec:
  hostNetwork: true
  dnsPolicy: ClusterFirstWithHostNet
  containers:
    - name: redis
      image: redis:latest
      ports:
        - name: web
          containerPort: 6379
          protocol: TCP
```
3. Run agent's inspect command to verify if an input is created `./elastic-agent -c ./elastic-agent.yml inspect output -o default`
4. Verify that an input is created and metadata are added properly via the processors, for example:
```
metricbeat:
  modules:
  - hosts:
    - 172.18.0.3:6379
    idle_timeout: 20s
    index: metrics-redis.info-default
    maxconn: 10
    meta:
      package:
        name: redis
        version: 0.3.6
    metricsets:
    - info
    module: redis
    name: redis
    network: tcp
    period: 10s
    processors:
    - add_fields:
        fields:
          id: cda93b8385201eaf725532e908e4a4c55895273d2a2cb57700b3cf48732b2de3
          image:
            name: redis:latest
          runtime: containerd
        target: container
    - add_fields:
        fields:
          container:
            name: redis
          labels:
            app: redis
            role: main
          namespace: default
          node:
            name: kind-control-plane
          pod:
            ip: 172.18.0.3
            name: redis
            uid: 365e8d21-a81d-44e2-acbf-95deb8a57730
        target: kubernetes
    - add_fields:
        fields:
          cluster:
            name: kind-kind
            url: https://127.0.0.1:50740
        target: orchestrator
    - add_fields:
        fields:
          dataset: redis.info
          namespace: default
          type: metrics
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.info
        target: event
    - add_fields:
        fields:
          id: 32806c0a-4f04-499c-9427-e6e24e5f6035
          snapshot: false
          version: 8.0.0
        target: elastic_agent
    - add_fields:
        fields:
          id: 32806c0a-4f04-499c-9427-e6e24e5f6035
        target: agent
```

### Using annotations (Note: annotations are not exposed to meta fields by default but are exposed in variable resolution mechanism)
Specify a condition based on an annotation like `condition: ${kubernetes.annotations.level} == 'production' AND ${kubernetes.container.port_name} == 'web'` and verify that input is created again. (note: tune Redis Pod manifest accordingly)


### Short-living pods
Define a log input with condition based on container's name:
```
      - name: container-log
        type: logfile
        use_output: default
        meta:
          package:
            name: log
            version: 0.4.6
        data_stream:
          namespace: default
        streams:
          - data_stream:
              dataset: generic
            symlinks: true
            condition: ${kubernetes.container.name} == 'hello'
            paths:
              - /var/log/containers/*${kubernetes.container.id}.log
```

Deploy elastic-agent and verify that it collects logs for the following short-living pod:
```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: mytarge
  labels:
    app: mytarget
spec:
  schedule: "*/1 * * * *"
  failedJobsHistoryLimit: 10
  successfulJobsHistoryLimit: 20
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            args:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```
![Screenshot 2021-09-15 at 3 56 18 PM](https://user-images.githubusercontent.com/11754898/133437309-5c033383-6fd9-43a0-89e1-84ae24d7aa63.png)


### Short living Init containers are covered
```
---
apiVersion: v1
kind: Pod
metadata:
  name: mytarget
  labels:
    app: test
spec:
  initContainers:
    - name: test-init
      image: ubuntu:latest
      command:
        - bash
        - -c
        - |
          #!/bin/bash
          echo "$(date): started the INIT process"
          sleep 10
          echo "$(date): sleeping 10 INIT seconds"
  containers:
    - name: test
      image: ubuntu:latest
      command:
        - bash
        - -c
        - |
          #!/bin/bash
          echo "$(date): started the process"

          while :
          do
                 echo "$(date): sleeping 5 seconds"
                 sleep 5
          done
```

### Ensure logs are captured even if Pod is marked for deletion
```
---
apiVersion: v1
kind: Pod
metadata:
  name: mytarget2
  labels:
    app: test
spec:
  containers:
    - name: test
      image: ubuntu:latest
      command:
        - bash
        - -c
        - |
          #!/bin/bash
          trap "{ echo '$(date): SIGTERM triggered'; sleep 3;echo '$(date): Bye bye'; sleep 5; echo '$(date): bye'; sleep 5; exit 1; }" SIGINT SIGTERM
          echo "$(date): started the process"

          while :
          do
                 echo "$(date): sleeping 5 seconds"
                 sleep 5
          done
```

## Related issues
- Closes https://github.com/elastic/beats/issues/24495, https://github.com/elastic/beats/issues/27019

## Notes for the reviewer
Most of the functionality is ported from `libbeat/autodiscover/providers/kubernetes` and `libbeat/common/kubernetes/metadata`.
